Add CI with link checker.

dblock opened the pull request:
### Description

Builds Jekyll site similarly to project-website, runs link checker, fails if a link is broken.

Link checker had some improvements, makes HEAD requests first, which is a lot faster.

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

kolchfa-aws commented on _plugins/link-checker.rb:
```diff
@@ -1,219 +1,376 @@
+  def self.verify(_site)
+    return unless @check_links
+
+    @base_url_matcher = %r{^#{@site.config["url"]}#{@site.baseurl}(/.*)$}.freeze
```
Could we add a check for the case where a writer forgets a slash after the base URL? For example, `{{site.url}}{{site.baseurl}}opensearch/supported-field-types/range/`.

dblock replied:
It's a bit tricky, let's defer this to a future PR.

kolchfa-aws commented on _plugins/link-checker.rb:
```diff
@@ -1,219 +1,376 @@
+    end
+
+    if !@failures.empty?
+      raise msg if @should_build_fatally
```
I think a gentler way would be to just print an error message and exit(1). Writers who are running a link checker locally may attribute an exception to something not working in the script. If we're raising an exception, I believe it should be raise(msg) because msg is a string.

dblock replied:
Actually this doesn't work. If you `exit(1)` Jekyll will display this as an error, instead of displaying the actual error.

```
/home/runner/work/documentation-website/documentation-website/_plugins/link-checker.rb:195:in `exit': exit (SystemExit)
	from /home/runner/work/documentation-website/documentation-website/_plugins/link-checker.rb:195:in `verify'
	from /home/runner/work/documentation-website/documentation-website/_plugins/link-checker.rb:134:in `block in init'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/hooks.rb:103:in `block in trigger'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/hooks.rb:102:in `each'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/hooks.rb:102:in `trigger'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/site.rb:234:in `write'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/site.rb:82:in `process'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/command.rb:28:in `process_site'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:65:in `build'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/commands/build.rb:36:in `process'
	from /home/runner/work/documentation-website/documentation-website/vendor/bundle/ruby/3.0.0/gems/jekyll-4.2.2/lib/jekyll/command.rb:91:in `block in process_with_graceful_fai
```

kolchfa-aws commented on _plugins/link-checker.rb:
```diff
@@ -1,219 +1,376 @@
+    if !@failures.empty?
+      raise msg if @should_build_fatally
+
+      Jekyll.logger.warn "\nLinkChecker: [Warning] #{msg}\n"
```
Can we print this in red to attract attention?
```suggestion
      Jekyll.logger.warn "\nLinkChecker: [Warning] #{msg}\n".red
```

kolchfa-aws reviewed (approved):
Thank you, dB! This will partly close #681. The only other improvement would be to add a broken image check.

kolchfa-aws commented on .github/workflows/jekyll-build.yml:
```diff
@@ -0,0 +1,16 @@
+name: Jekyll Build Verification
+
+on: [pull_request]
```
Maybe we should also run an all/external link checker on a cron schedule (every week?) because sometimes external links change over time. We can create an issue if it fails. Also, does this run a link checker on all files on every PR? If we add a run on a schedule, then I think it may be sufficient to just check changed files on every PR.

dblock commented:
In https://github.com/opensearch-project/project-website/pull/1470 we changed the link checker to run on push to main only (not in PRs) and autocut a ticket. Would you prefer that to the implementation that checks on every PR, @kolchfa-aws?

kolchfa-aws commented:
@dblock I think the best way would be to run an internal link checker on every PR, and run an all link checker (primarily for the purpose of identifying external links) on a cron schedule once a week and create an issue if there are failures. Alternatively, we can run an all link checker on every PR. Also, we are running an internal link checker locally every time we build, so could you please make a change to the build.sh file to include an internal link checker env variable?
```JEKYLL_LINK_CHECKER=internal bundle exec jekyll serve --host localhost --port 4000 --incremental --livereload --open-url --trace``` Thanks!

dblock commented:
@kolchfa-aws Updated!

- Updated to use https://github.com/dblock/ruby-link-checker/ which is many times faster than the implementation I copied from project-website.
- Updated build.sh and PR to run local link checker.
- Rather than run on cron, split external link checker to run on every push, and open a GitHub issue on failure. Passing checker in https://github.com/dblock/documentation-website/actions/runs/4598675096.
- Did not change the way link checker fails as Jekyll treats `exit` as an exception.
- Corrected more broken links.

kolchfa-aws commented:
@dblock Thanks so much! Is there a way to run the local build in quiet mode so it only displays errors and not info? If not, that's fine because error messages come last. It would be nice to exit without an exception, but if it's not possible, it's fine as well.

cwillum commented on _observing-your-data/ad/index.md:
```diff
@@ -14,7 +14,7 @@ An anomaly in OpenSearch is any unusual behavior change in your time-series data
 It can be challenging to discover anomalies using conventional methods such as creating visualizations and dashboards. You could configure an alert based on a static threshold, but this requires prior domain knowledge and isn't adaptive to data that exhibits organic growth or seasonal behavior.
 
-Anomaly detection  automatically detects anomalies in your OpenSearch data in near real-time using the Random Cut Forest (RCF) algorithm. RCF is an unsupervised machine learning algorithm that models a sketch of your incoming data stream to compute an `anomaly grade` and `confidence score` value for each incoming data point. These values are used to differentiate an anomaly from normal variations ...
+Anomaly detection  automatically detects anomalies in your OpenSearch data in near real-time using the Random Cut Forest (RCF) algorithm. RCF is an unsupervised machine learning algorithm that models a sketch of your incoming data stream to compute an `anomaly grade` and `confidence score` value for each incoming data point. These values are used to differentiate an anomaly from normal variations ...
```
extra space after "detection"?

cwillum commented on _plugins/link-checker.rb:
```diff
@@ -1,219 +1,283 @@
+
+  ##
+  # Defines the priority of the plugin
+  # The hooks are registered with a very low priority to make sure they runs after any content modifying hook
```
"... make sure they run after ..." verb agreement
Not sure if it's necessary to make proper sentences above here and add periods. Just notes, I suppose.

dblock replied:
I won't change them all just to match the code in project-website repo for now.

cwillum commented on _tools/logstash/read-from-opensearch.md:
```diff
@@ -48,4 +48,4 @@ Like the output plugin, after adding your configuration to the `pipeline.conf` f
 Adding `stdout{}` to the `output{}` section of your `pipeline.conf` file prints the query results to the console. 
 
-To reindex the data into an OpenSearch domain, add the destination domain configuration in the `output{}` section like shown [here](https://opensearch.org/docs/latest/clients/logstash/ship-to-opensearch/#opensearch-output-plugin).
+To reindex the data into an OpenSearch domain, add the destination domain configuration in the `output{}` section like shown [here](https://opensearch.org/docs/latest/tools/logstash/index/).
```
"... section, as shown in the Logstash documentation."

dblock replied:
Going to skip changing this, it's an existing content on the website and I'm only fixing links.

cwillum reviewed (approved):
LGTM. just nit picking comments.

dtaivpp commented:
@kolchfa-aws would it be helpful if it ran locally every time before you tried to push? I've thought about adding local commit hooks that could run things like this automatically when you push.

kolchfa-aws commented:
@dtaivpp We're running a build script locally when we build the site that checks internal links (build.sh). I think it accomplishes the same thing.

kolchfa-aws commented:
@dblock Could we run fatally on a pull request? We would like the PR to fail if it has broken links.

dblock commented:
> @dblock Could we run fatally on a pull request? We would like the PR to fail if it has broken links.

We could, but then it would be different from https://github.com/opensearch-project/project-website/pull/1470, which had the exact opposite opinion via @krisfreedain :) What would you like to do? 

The argument against breaking on PR is that it takes several minutes to run the link checker, and most broken links will be outside of the scope of the PR being made IMO (some existing website, partner, etc., disappearing). With this change as is you'll get a GitHub issue on every push that some link is broken, so it can be dealt with separately.

dblock commented:
> @dblock Thanks so much! Is there a way to run the local build in quiet mode so it only displays errors and not info? If not, that's fine because error messages come last. It would be nice to exit without an exception, but if it's not possible, it's fine as well.

Not easily. Let's consider this as future improvements.

kolchfa-aws commented:
@dblock After talking it over with the team, we'd like to try the following if possible:

- On every PR, check internal links and don't allow merging until the internal links are fixed.
- On a cron schedule, once a week, check all links and create an issue with the links that are broken.

I realize that this is different from what the project site is doing, but I think it's ok. Our main goal is to enforce internal link fixing, where we see the most problems. If the run takes several minutes, it's ok as well. And if we don't allow merging without link fixing, then theoretically the only broken links in a given PR should come only from the changes in that PR. If we do that, can we then the link checker rerun automatically when there is a push to the same PR that fixes the links?

Also, if we could annotate the PR with the list of broken links and the file names where the broken links exist, that would be ideal. Thank you!

krisfreedain commented:
> > @dblock Could we run fatally on a pull request? We would like the PR to fail if it has broken links.
> 
> We could, but then it would be different from [opensearch-project/project-website#1470](https://github.com/opensearch-project/project-website/pull/1470), which had the exact opposite opinion via @krisfreedain, and different from what you suggested in [#3584 (comment)](https://github.com/opensearch-project/documentation-website/pull/3584#issuecomment-1491011274) :) What would you like to do?
> 
> The argument against breaking on PR is that it takes several minutes to run the link checker, and most broken links will be outside of the scope of the PR being made IMO (some existing website, partner, etc., disappearing). With this change as is you'll get a GitHub issue on every push that some link is broken, so it can be dealt with separately.

yeah - on the project-website, I'm going to file an issue to have it removed from the build process completely and just be something similar to @kolchfa-aws's feedback of "On a cron schedule, once a week, check all links and create an issue with the links that are broken."

dblock commented:
> > > @dblock Could we run fatally on a pull request? We would like the PR to fail if it has broken links.
> > 
> > 
> > We could, but then it would be different from [opensearch-project/project-website#1470](https://github.com/opensearch-project/project-website/pull/1470), which had the exact opposite opinion via @krisfreedain, and different from what you suggested in [#3584 (comment)](https://github.com/opensearch-project/documentation-website/pull/3584#issuecomment-1491011274) :) What would you like to do?
> > The argument against breaking on PR is that it takes several minutes to run the link checker, and most broken links will be outside of the scope of the PR being made IMO (some existing website, partner, etc., disappearing). With this change as is you'll get a GitHub issue on every push that some link is broken, so it can be dealt with separately.
> 
> yeah - on the project-website, I've filed an issue to have it removed from the build process completely ([opensearch-project/project-website#1501](https://github.com/opensearch-project/project-website/issues/1501)) and just be something similar to @kolchfa-aws's feedback of "On a cron schedule, once a week, check all links and create an issue with the links that are broken."

For project-website I believe I have fixed the false positive problem in https://github.com/opensearch-project/project-website/issues/1501#issuecomment-1496285626.  If that doesn't help we can change to cron.

dblock commented:
@kolchfa-aws merge?

kolchfa-aws commented:
@dblock We would prefer to run an all link checker on cron rather than on push. Since internal links will be caught on every PR, this leaves external links only, which don't change as often, so it should be caught early enough. 

If you could change an all link checker to run on a cron schedule, we can definitely merge the PR and see how it goes. We can always go back to checking on every push.

kolchfa-aws commented:
Thank you, @dblock!

dblock commented:
@kolchfa-aws do you need me to backport some of these?

dblock commented:
Link checker ran successfully in https://github.com/opensearch-project/documentation-website/actions/runs/4618022749 fyi.

kolchfa-aws commented:
@dblock I already saw. Nice!
